### PR TITLE
Softly check whether the gpu is avalid and whether the index is availab…

### PR DESCRIPTION
**Pull request by Presburger**
/kind feature
issue: #485 
issue: #57

### CODE REVIEW

File: src/common/factory.cc [hunk 1]
```diff
@@ -11,20 +11,45 @@
...
-    LOG_KNOWHERE_INFO_ << "use key " << key << " to create knowhere index " << name << " with version " << version;
+    if (func_mapping_.find(key) == func_mapping_.end()) {
+        LOG_KNOWHERE_INFO_ << "use key " << key << " to create knowhere index " << name << " with version " << version;
+        return expected<Index<IndexNode>>::Err(Status::invalid_index_error, "index not supported");
+    }
     auto fun_map_v = (FunMapValue<Index<IndexNode>>*)(func_mapping_[key].get());
+
+#ifdef KNOWHERE_WITH_RAFT
+
+    if (name == "GPU_RAFT_BRUTE_FORCE" || name == "GPU_BRUTE_FORCE" || name == "GPU_RAFT_CAGRA" ||
```
Comment (foxspy) [inline src/common/factory.cc:37]: move this condition prediction into util class

**Comment by Presburger** [comment]
kind/feature

**Comment by Presburger** [comment]
issue: #57

**Review by liliu-z — APPROVED** [review]
/lgtm

**Pull request merged**